File: thingseong/두 용액.py
import sys

n = int(sys.stdin.readline())
A = list(map(int, sys.stdin.readline().split()))
A = sorted(A)

l, r = 0, n-1
m = None
res = None

while(l < r):
    tmp = abs(A[l] + A[r])
    if m == None:
        m = tmp
        res = (A[l], A[r])
    elif tmp < m:
        m = tmp
        res = (A[l], A[r])
        

    if A[l] + A[r] == 0:
        res = (A[l], A[r])
        break
    elif A[l] + A[r] < 0:
        l = l + 1
    elif A[l] + A[r] > 0:
        r = r - 1


print(res[0], res[1])

File: vgihan/baekjoon/두용액.js
const readline = require("readline");
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let n;

rl.on("line", function (line) {
  if (!n) {
    n = parseInt(line);
    return;
  }
  console.log(solution(line.split(" ").map((v) => parseInt(v))).join(" "));
  rl.close();
}).on("close", function () {
  process.exit();
});

function solution(inputs) {
  const liquids = [...inputs.sort((a, b) => a - b)];
  let i = 0;
  let j = liquids.length - 1;
  let min = 10000000000;
  let result;

  while (i < j) {
    if (liquids[i] >= 0) j = i + 1;
    const sum = liquids[i] + liquids[j];
    if (Math.abs(sum) <= Math.abs(min)) {
      result = [liquids[i], liquids[j]];
      min = Math.abs(sum);
    }
    if (sum <= 0) {
      i++;
      continue;
    }
    if (sum > 0) {
      j--;
      continue;
    }
  }

  return result;
}
